sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -501,6 +501,10 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
         case LessThanOrEqual(l, r) => GreaterThan(l, r)
         // not(not(e))  =>  e
         case Not(e) => e
+        // De Morgan's law: !(a || b) => !a && !b
```

Comment:
To the `BooleanSimplification`, this probably not an optimization, isn't it? Probably we'd better to move this logic into `predicates.scala`, which used only for the cases for the join predicate push down.